encrypted-media-respec.html

```diff
@@ -2101,6 +2101,30 @@
           <p>The following steps are run:</p>
           <ol>
             <li><p>Let the <var>session</var> be the associated <a>MediaKeySession</a> object.</p></li>
+            <li><p>Run the <a def-id="destroy-keys-and-licenses-algorithm"></a> algorithm.</p></li>
+            <li><p>Run the <a def-id="update-key-statuses-algorithm"></a> algorithm on the <var>session</var>, providing an empty sequence.</p></li>
+            <li><p>Run the <a def-id="update-expiration-algorithm"></a> algorithm on the <var>session</var>, providing <code>NaN</code>.</p></li>
+            <li><p>Let <var>promise</var> be the <a def-id="closed"></a> attribute of the <var>session</var>.</p></li>
+            <li><p>Resolve <var>promise</var>.</p></li>
+          </ol>
+        </section>
+        <section id="destroy-keys">
+          <h4>Destroy Keys and Licenses</h4>
+          <p>
+            The Destroy Keys and Licenses algorithm is run when the CDM needs to destroy all the keys and licenses associated with a session that
+            have not been explicitly stored.
+          </p>
+          <p>
+            For sessions of type <a def-id="temporary-session"></a> and <a def-id="persistent-usage-record-session"></a>, this algorithm MUST be
+            executed when the <a>MediaKeySession</a> object becomes inaccessible to the page, if the <a def-id="session-close-algorithm"></a>
+            algorithm has not already been run.
+          </p>
+          <p class="note">
+            This is necessary to ensure that the keys and licenses associated with the session are correctly destroyed when the session ends.
+          </p>
+          <p>The following steps are run:</p>
+          <ol>
+            <li><p>Let the <var>session</var> be the associated <a>MediaKeySession</a> object.</p></li>
```

Comment:
This seems to be the main issue, so I'll put all my comments here.

There are several examples where we are not all that clear about exactly how the UA and CDM interact to achieve the steps of our algorithms. The "is run when" phrase is a red flag for this issue, as you say. For this algorithm and indeed Session Close there are likely two distinct cases in any given implementation:
(a) The UA initiates the action, explicitly telling the CDM to either close a session or destroy keys associated with the session (e.g. `close()`, `MediaKeySession` destroyed, `remove()` if we include `persistent` sessions)
(b) The CDM initiates the action, informing the UA that it has either closed a session or destroyed keys associated with the session (resources lost by the CDM, other CDM-specific reasons).

I think the confusion here arises from trying to cover both cases in one piece of text.

For "Session Close", we could rename it to "Session Closed" and describe explicitly that these are the steps the UA performs after the session is closed and that they are triggered either explicitly in other algorithms of after an indication from the CDM that it has initiated closure of the session.

For "Destroy Keys.." we would split this into one part which is triggered by the UA from `remove()` (for `persistent` session) and when `MediaKeySession` is destroyed for `temporary` and `persistent-usage-record` and a second part ('Keys Destroyed') which is triggered when the CDM initiates destruction of the keys itself.